frontend/src/metabase/dashboard/components/DashCard/DashCardParameterMapper/DashCardCardParameterMapper.jsx

```diff
@@ -113,6 +111,33 @@ export function DashCardCardParameterMapper({
   const virtualCardType = getVirtualCardType(dashcard);
   const isNative = isNativeDashCard(dashcard);
 
+  let selectedMappingOption;
+  if (isVirtual || isAction || isNative) {
+    selectedMappingOption = target
+      ? mappingOptions.find(option =>
+          _.isEqual(normalize(option.target), normalize(target)),
+        )
+      : undefined;
+  } else {
+    const stageIndex = -1;
+    const columns = Lib.visibleColumns(question.query(), stageIndex);
+
+    selectedMappingOption = target
+      ? mappingOptions.find(mappingOption => {
+          const [index1, index2] = target
+            ? Lib.findColumnIndexesFromLegacyRefs(
```

Comment:
It seems we have O(n^2) here because for every `mappingOption` we call `findColumnIndexesFromLegacyRefs` that iterates on each `column`. Can it be improved? Maybe something like this:

```
const mappingColumnIndexes = Lib.findColumnIndexesFromLegacyRefs(
  query,
  stageIndex,
  columns,
  mappingOptions.map(({ target }) => target[1]), 
);

const [selectedColumnIndex] = Lib.findColumnIndexesFromLegacyRefs(
  query,
  stageIndex,
  columns,
  [target[1]],
);

if (selectedColumnIndex >= 0) {
  const mappingIndex = mappingColumnIndexes.indexOf(selectedColumnIndex);
  if (mappingIndex >= 0) {
    return mappingOptions[mappingIndex];
  }
}

return undefined;
```

Reply:
fixed here [480e5e3](https://github.com/metabase/metabase/pull/38545/commits/480e5e3220108fac3861e2ae2f47a2319f9c9839)